kuma/redirects/redirects.py

```diff
@@ -762,7 +775,604 @@ def locale_redirect(pattern, to, prepend_locale=True, **kwargs):
              permanent=False),
 ]
 
-redirectpatterns = scl3_redirectpatterns + [
+zone_redirectpatterns = [
+    # The redirects for the case when the locale is not specified
+    # for a zone. It must be handled here, since for the cases when
+    # LocaleMiddleware handles the 404 response and redirects to the
+    # proper locale, it fails because the path is considered invalid.
+    redirect(
+        r'^(?:docs/)?Add-ons(?:/?|(?P<sub_path>[/$].+))$',
```

Comment:
Is it possible to write the things in logic?
something like
```python
locales = ['en-US', 'bn-BD']
redirects = []
for locale in locales:
    r = redirect(
        r'^{locale}/(?:docs/)?Add-ons(?:/?|(?P<sub_path>[/$].+))$'.format(locale=locale),
        get_sub_path_handler(u'/{locale}/docs/Mozilla/Add-ons'.format(locale=locale)),
        permanent=False,
        decorators=shared_cache_control_for_zones),
    redirects.append(r)
```

Reply:
I agree but would do it differently, to capture all the zone redirect patterns in the data structure:

```python
zone_redirects = (
    ('Add-ons', 'Mozilla/Add-ons', ('en-US', 'bn-BD')),
)
zone_pattern = r'^{locale}/(?:docs/)?{zone_root}(?:/?|(?P<sub_path>[/$].+))$'
sub_path_fmt = u'/{locale}/docs/{wiki_slug}'

zone_redirectpatterns = []
for zone_root, wiki_slug, locales in zone_redirects:
    for locale in locales:
        pattern = zone_pattern.format(locale=locale, zone_root=zone_root)
        sub_path = sub_path_fmt.format(locale=locale, wiki_slug=wiki_slug)
        zone_redirectpatterns.append(redirect(
            pattern,
            get_sub_path_handler(sub_path),
            permanent=False,
            decorators=shared_cache_control_for_zones))
```

There is a benefit to the verbose style in this PR, that it makes it easier to confirm that all the zones were migrated.  I'm OK shipping this PR as written and removing the repetition later, or making this change in this PR.

Reply:
Nice, I like it. You haven't left any fun for me! 😀